datalad/interface/diff.py

```diff
@@ -0,0 +1,276 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Plumbing command for reporting changes in datasets"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+import stat
+from os.path import join as opj
+from os.path import curdir
+from os.path import relpath
+
+
+from datalad.interface.annotate_paths import AnnotatePaths
+from datalad.interface.annotate_paths import annotated2content_by_ds
+from datalad.interface.base import Interface
+from datalad.interface.utils import eval_results
+from datalad.interface.utils import build_doc
+from datalad.support.constraints import EnsureNone
+from datalad.support.constraints import EnsureStr
+from datalad.support.constraints import EnsureChoice
+from datalad.support.exceptions import CommandError
+from datalad.support.exceptions import InsufficientArgumentsError
+from datalad.support.param import Parameter
+from datalad.interface.common_opts import recursion_flag
+from datalad.interface.common_opts import recursion_limit
+from datalad.cmd import GitRunner
+
+from datalad.distribution.dataset import EnsureDataset
+from datalad.distribution.dataset import datasetmethod
+
+lgr = logging.getLogger('datalad.interface.diff')
+
+
+# from Git docs
+state_map = {
+    'A': 'added',
+    'C': 'copied',
+    'D': 'deleted',
+    'M': 'modified',
+    'R': 'renamed',
+    'T': 'typechange',
+    'U': 'unmerged',
+    'X': 'unknown_potentialbug',
+}
+
+
+def _translate_status(label, ap):
+    if label[0] in ('C', 'R'):
+        ap['perc_similarity'] = float(label[1:])
+        label = label[0]
+    ap['state'] = state_map[label]
+
+
+def _translate_type(mode, ap, prop):
+    if mode == 0:
+        ap[prop] = None
+    elif mode == stat.S_IFDIR | stat.S_IFLNK:
+        ap[prop] = 'dataset'
+    elif stat.S_ISDIR(mode):
+        # not sure if this can happen at all
+        ap[prop] = 'directory'
+    else:
+        ap[prop] = 'file'
+
+
+def _parse_git_diff(dspath, diff_thingie=None, paths=None,
```

Comment:
And DOC: Not sure, what a `diff_thingie` is supposed to be. Also: a `path` seems to be considered only if it is a raw input, right?


Generally, one can find out what those helpers are for, but it requires to go through the entire code of the helper itself and at least some usages of them to figure it out. This might be okayish for a complete review (still makes it harder), but within a week one needs to do it again to recall ... ;-)